### PR TITLE
Add New Tasking I/F and rework GPIO, UART, etc to support it

### PR DESCRIPTION
**Pull request by TerryE**
As with the last commit this rolls up the follwowing, but include the various
review comments on the PR.

-   **Documentation changes**. I've added the taks FAQ as a stub new Extension
developer FAQ, and split the old FAQ into a Lua Developer FAQ and a Hardware
FAQ.

-   **Tasking I/F**.  New `app/task/Makefile`, `app/task/task.c`,
`app/include/task/task.h` and `app/Makefile` as per previous commit.  Cascade
changes to `app/driver/uart.c`, `app/include/driver/uart.h`,
`app/user/user_main.c` and `app/modules/node.c`

-   **GPIO Rework** to `app/modules/gpio.c` and `pin_map.[hc]`, `platform.[hc]`
in `app/platform`

-   **Other Optimisations** Move the `platform_*_exists()` from
`app/platform/common.c` to static inline declarations in `platform.h` as
this generates faster, smaller code. Move lgc.a routines out of iram0.

### CODE REVIEW

**Comment by TerryE** [comment]
Please refer to #1023 for previous comments and feedback.

File: app/modules/gpio.c [hunk 1]
```diff
@@ -16,139 +16,115 @@
...
-  if (lua_type(L, 3) == LUA_TFUNCTION || lua_type(L, 3) == LUA_TLIGHTFUNCTION){
+  unsigned pin = luaL_checkinteger( L, 1 );
+  static const char * const opts[] = {"", "up", "down", "both", "low", "high"};
+  static const int opts_type[] = {
+    GPIO_PIN_INTR_DISABLE, GPIO_PIN_INTR_POSEDGE, GPIO_PIN_INTR_NEGEDGE,
+    GPIO_PIN_INTR_ANYEDGE, GPIO_PIN_INTR_LOLEVEL, GPIO_PIN_INTR_HILEVEL
+    };
+  int type = opts_type[luaL_checkoption(L, 2, "", opts)];
+
+  luaL_argcheck(L, platform_gpio_exists(pin) && pin>0, 1, "Invalid interrupt pin");
```
Comment (jmattsson) [inline app/modules/gpio.c]: Thanks for the `platform_gpio_exists()`, but why the `pin>0`? If it's due to pin 0 not being interrupt capable, I'd prefer that abstracted out behind a `platform_gpio_intr_capable(pin)` instead.

File: docs/en/hardware-faq.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+menu tab), or 
+
+-   Use the latest Espressif's flash tool -- see [this Espressif forum
+topic](http://bbs.espressif.com/viewtopic.php?f=5&t=433) (without auto download
+support). Use DIO mode and 32M flash size option, and flash latest firmware to
+0x00000. Before flashing firmware, remember to hold FLASH button, and press RST
+button once. Note that the new NodeMCU our firmware download tool, when
+released, will be capable of flashing firmware automatically without any button
+presses.
+
```
Comment (jmattsson) [inline docs/en/hardware-faq.md]: This sentence reads funny. Since it's speaking of a hypothetical future, maybe leave it out for now?

**Comment by jmattsson** [comment]
:+1: Unless someone raises something serious, I'll merge this Real Soon Now(tm).

File: app/modules/gpio.c (hunk 1 above)
Comment (TerryE) [inline app/modules/gpio.c]: OK, but there's a load of other tidy up then needs doing. Let's start keeping a list, but it's not worth delaying this commit for.

File: docs/en/hardware-faq.md (hunk 2 above)
Comment (TerryE) [inline docs/en/hardware-faq.md]: Yup chinenglish. But we've got to tidy up the docs. Second pass.

**Comment by TerryE** [comment]
Please. I w and to move on to clearing off my pending list :smile:

**Comment by jmattsson** [comment]
3...2..1...merged! I think this is probably the most in-depth reviewed PR we've had already.

**Pull request merged**